astropy/wcs/src/wcslib_prjprm_wrap.c

```diff
@@ -0,0 +1,1024 @@
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+
+/***************************************************************************
+ * PyPrjprm methods                                                        *
+ ***************************************************************************/
+
+static PyObject* PyPrjprm_new(PyTypeObject* type, PyObject* args, PyObject* kwds)
+{
+    PyPrjprm* self;
+    self = (PyPrjprm*)type->tp_alloc(type, 0);
+    if (self == NULL) return NULL;
+    self->owner = NULL;
+    self->x = NULL;
+    self->prefcount = NULL;
+    if ((self->x = calloc(1, sizeof(struct prjprm))) == 0x0) {
+        PyErr_SetString(PyExc_MemoryError, "Could not allocate memory.");
+        return NULL;
+    }
+    if ((self->prefcount = (int*) malloc(sizeof(int))) == 0x0) {
+        PyErr_SetString(PyExc_MemoryError, "Could not allocate memory.");
+        free(self->x);
+        return NULL;
+    }
+    if (wcslib_prj_to_python_exc(prjini(self->x)))
+    {
+        free(self->x);
+        free(self->prefcount);
+        return NULL;
+    }
+    *(self->prefcount) = 1;
+    return (PyObject*)self;
+}
+
+
+static int PyPrjprm_traverse(PyPrjprm* self, visitproc visit, void *arg)
+{
+    Py_VISIT(self->owner);
+    return 0;
+}
+
+
+static int PyPrjprm_clear(PyPrjprm* self)
+{
+    Py_CLEAR(self->owner);
+    return 0;
+}
+
+
+static void PyPrjprm_dealloc(PyPrjprm* self)
+{
+    PyPrjprm_clear(self);
+    if (!self->owner) {
+        wcslib_prj_to_python_exc(prjfree(self->x));
+        if (self->prefcount && (--(*self->prefcount)) == 0) {
+            free(self->x);
+            free(self->prefcount);
+        }
+    }
+    Py_TYPE(self)->tp_free((PyObject*)self);
+}
+
+
+PyPrjprm* PyPrjprm_cnew(PyObject* celprm_obj, struct prjprm* x, int* prefcount)
+{
+    PyPrjprm* self;
+    self = (PyPrjprm*)(&PyPrjprmType)->tp_alloc(&PyPrjprmType, 0);
+    if (self == NULL) return NULL;
+    self->x = x;
+    Py_XINCREF(celprm_obj);
+    self->owner = celprm_obj;
+    self->prefcount = prefcount;
+    if (prefcount) (*prefcount)++;
+    return self;
+}
+
+
+static PyObject* PyPrjprm_copy(PyPrjprm* self)
+{
+    PyPrjprm* copy = NULL;
+    copy = PyPrjprm_cnew(self->owner, self->x, self->prefcount);
+    if (copy == NULL) return NULL;
+    return (PyObject*)copy;
+}
+
+
+static PyObject* PyPrjprm_deepcopy(PyPrjprm* self)
+{
+    PyPrjprm* copy = PyPrjprm_new(&PyPrjprmType, NULL, NULL);
+    if (copy == NULL) return NULL;
+
+    memcpy(copy->x, self->x, sizeof(struct prjprm));
+    copy->x->err = NULL;
+    return (PyObject*)copy;
+}
+
+
+static PyObject* PyPrjprm___str__(PyPrjprm* self)
+{
+    wcsprintf_set(NULL);
+    if (wcslib_prj_to_python_exc(prjprt(self->x))) {
+        return NULL;
+    }
+    return PyUnicode_FromString(wcsprintf_buf());
+}
+
+static int PyPrjprm_cset(PyPrjprm* self)
+{
+    if (wcslib_prj_to_python_exc(prjset(self->x))) {
+        return -1;
+    }
+    return 0;
+}
+
+
+static PyObject* PyPrjprm_set(PyPrjprm* self)
```

Comment:
Yes, this is (may be) needed. For example, one could want to modify some values in `prjprm.pv` or `phi0` (etc.) and ask `set` to re-compute the rest of the values. Without `set()` one has to call a coordinate transformation in order to update values in `prjprm`. In the end, if `set()` is not needed when using `prjprm` why is WCSLIB proving it?

Reply:
> Note that this routine need not be called directly; it will be invoked by prjx2s() and prjs2x() if prj.flag is anything other than a predefined magic value.

Most importantly, `prjx2s` and `prjs2x` pointers are NULL in the `prjini()` routine:
```C
  prj->prjx2s = 0x0;
  prj->prjs2x = 0x0;
```
and hence `prjx2s()` and `prjs2x()` are not available to "set" `prjprm` at that time. First, one has to call `set` at least once _after_ `code` is set to set `prjx2s` and `prjs2x` pointers to the correct function. In principle, this could be done in class initializer and require that a code be provided but this would complicate the other PR and I do not see what is wrong with giving control to the user/caller.

Reply:
👍 so "need not" != "should not".  Sounds good.